equational_theories/MemoFinOp.lean

```diff
@@ -0,0 +1,67 @@
+import Lean
+
+/-!
+This file defines a macro `memoFinOp` that takes a function `f : Fin n → Fin n → Fin n`, at
+elaboration time calculates its table of values, encodes it into a single `Nat`, and returns a term
+that implements the operation via lookup in this table.
+-/
+
+open Lean Meta Elab  Term
+
+def opOfTable {n : Nat} (t : Nat) (a : Fin n) (b : Fin n) : Fin n :=
+  let i := a.val * n + b.val
+  let r := (t / n^i) % n
+  ⟨r, Nat.mod_lt _ (by exact Fin.pos a)⟩
+
+/-- `enum n` is the array of all elements of `Fin n` in order -/
+private def Fin.enum (n) : Array (Fin n) := Array.ofFn id
```

Comment:
```suggestion
private def Array.finRange (n) : Array (Fin n) := Array.ofFn id
```
to match `List.finRange`

Reply:
It's a copy of Fin.enum from Batteries.Data.Fin.Basic. And it's private… :-)

Reply:
Ah, I mistakenly thought leanprover-community/batteries#899 had been merged.